admin-dev/themes/new-theme/js/components/modal/progress-modal.ts

```diff
@@ -0,0 +1,421 @@
+
+  body!: HTMLElement;
+
+  closeIcon!: HTMLButtonElement;
+
+  header!: HTMLElement;
+
+  message!: HTMLElement;
+
+  constructor(params: ProgressModalParams) {
+    this.buildModalContainer(params);
+  }
+
+  buildModalContainer(params: ProgressModalParams): void {
+    this.container = document.createElement('div');
+    this.container.classList.add('modal', 'fade');
+    this.container.id = params.id;
+
+    // Modal dialog element
+    this.dialog = document.createElement('div');
+    this.dialog.classList.add('modal-dialog');
+    if (params.dialogStyle) {
+      Object.keys(params.dialogStyle).forEach((key: string) => {
+        // @ts-ignore
+        this.dialog.style[key] = params.dialogStyle[key];
+      });
+    }
+
+    this.container.appendChild(this.dialog);
+  }
+}
+
+export class ProgressModalContainer {
+  footer!: HTMLElement;
+
+  container!: HTMLElement;
+
+  dialog!: HTMLElement;
+
+  content!: HTMLElement;
+
+  message!: HTMLElement;
+
+  header!: HTMLElement;
+
+  title?: HTMLElement;
+
+  body!: HTMLElement;
+
+  stopProcessingButton!: HTMLElement;
+
+  switchToErrorButton!: HTMLElement;
+
+  progressDone!: HTMLElement;
+
+  progressPercentage!: HTMLElement;
+
+  errorCount!: HTMLElement;
+
+  constructor(params: ProgressModalParams) {
+    this.buildModalContainer(params);
+  }
+
+  buildModalContainer(params: ProgressModalParams): void {
+    // Modal content element
+    this.content = document.createElement('div');
+    this.content.classList.add('modal-content');
+
+    // Modal message element
+    this.message = document.createElement('p');
+    this.message.classList.add('modal-message');
+
+    // Modal header element
+    this.header = document.createElement('div');
+    this.header.classList.add('modal-header');
+
+    params.modalTitle = params.modalTitle.replace('%d', params.total.toString());
+
+    this.progressPercentage = document.createElement('div');
+    this.progressPercentage.classList.add(
+      'float-right',
+      'progress-percent',
+    );
+    this.progressPercentage.append('0%');
+
+    this.body = document.createElement('div');
+    this.body.classList.add('modal-body', 'text-left', 'font-weight-normal');
+
+    this.body.append(this.progressPercentage);
+    this.body.append(this.buildProgressBar());
+
+    this.switchToErrorButton = document.createElement('div');
+    this.switchToErrorButton.classList.add(
+      'switch-to-errors-button',
+      'alert',
+      'alert-warning'
+    );
+
+    this.errorCount = document.createElement('span');
+    this.errorCount.classList.add(
+      'modal-error-count'
+    )
+    this.errorCount.innerHTML = '0';
+
+    let errorLogString = 'View %d error logs';
+    this.switchToErrorButton.innerHTML = errorLogString.replace('%d', this.errorCount.outerHTML);
+
+    this.header.append(this.switchToErrorButton);
+
+    let errorContainer = document.createElement('div');
+    errorContainer.classList.add('modal-error-container-single');
```

Comment:
Not sure this class is needed You don't need it for JS nor CSS for now (unless you plan to use it for CSS styling?)

Reply:
And I think you need to store this in `this.errorContainer` because when an error is added the progress modal is supposed to display the last error so in the `addError` function you will need to do something like this:

```js
this.progressModal.errorContainer.innerHTML = errorContent.outerHTML;
```